src/thermo/ThermoPhase.cpp

```diff
@@ -689,6 +689,617 @@ void ThermoPhase::setState_SPorSV(double Starget, double p,
+{
+    return getEquivalenceRatio_X(
+        parseCompString(fuelComp.find(":") != std::string::npos ? fuelComp : fuelComp+":1.0"),
+        parseCompString(oxComp.find(":") != std::string::npos ? oxComp : oxComp+":1.0"));
+}
+
+double ThermoPhase::getEquivalenceRatio_X(const vector_fp& fuelComp,
+                                          const vector_fp& oxComp) const
+{
+    return getEquivalenceRatio_X(fuelComp.data(), oxComp.data());
+}
+
+double ThermoPhase::getEquivalenceRatio_X(const double* fuelComp,
+                                          const double* oxComp) const
+{
+    auto Z = getMixtureFraction_X(fuelComp, oxComp);
+  
+    if (Z == 0.0)
+        return 0.0; // pure oxidizer
+
+    if (Z == 1.0) // pure fuel
+        return std::numeric_limits<double>::infinity();
+
+    auto stoichAirFuelRatio = getStoichAirFuelRatio_X(fuelComp, oxComp); 
+    
+    auto Xf_st = 1.0/(1.0+stoichAirFuelRatio);
+
+    auto sum_f = std::accumulate(fuelComp, fuelComp+m_kk, 0.0);
+    auto sum_o = std::accumulate(oxComp, oxComp+m_kk, 0.0);
+
+    auto mmw_st = 0.0;
+    for (size_t i=0; i!=m_kk; ++i)
+        mmw_st += molecularWeights()[i] * (Xf_st * fuelComp[i]/sum_f
+                  + (1.0-Xf_st) * oxComp[i]/sum_o);
+
+    auto mmw_f = std::inner_product(fuelComp, fuelComp+m_kk,
+                                    molecularWeights().data(), 0.0) / sum_f;
+
+    auto Z_st = Xf_st * mmw_f/mmw_st;
+
+    return std::max( Z / (1.0 - Z) * (1.0 - Z_st) / Z_st, 0.0);
+}
+
+double ThermoPhase::getEquivalenceRatio_Y(const compositionMap& fuelComp,
+                                          const compositionMap& oxComp) const
+{
+    vector_fp Yf(getCompositionFromMap(fuelComp));
+    vector_fp Yo(getCompositionFromMap(oxComp));
+    return getEquivalenceRatio_Y(Yf.data(), Yo.data());
+}
+
+double ThermoPhase::getEquivalenceRatio_Y(const std::string& fuelComp,
+                                          const std::string& oxComp) const
+{
+    return getEquivalenceRatio_Y(
+        parseCompString(fuelComp.find(":") != std::string::npos ? fuelComp : fuelComp+":1.0"),
+        parseCompString(oxComp.find(":") != std::string::npos ? oxComp : oxComp+":1.0"));
+}
+
+double ThermoPhase::getEquivalenceRatio_Y(const vector_fp& fuelComp,
+                                          const vector_fp& oxComp) const
+{
+    return getEquivalenceRatio_Y(fuelComp.data(), oxComp.data());
+}
+
+double ThermoPhase::getEquivalenceRatio_Y(const double* fuelComp,
+                                          const double* oxComp) const
+{
+    vector_fp X(2*m_kk); // mole fractions of fuel and oxidizer
+
+    // mean molecular weights of fuel and oxidizer
+    auto rmmw_f = std::inner_product(fuelComp, fuelComp+m_kk, molecularWeights().data(),
+            0.0, std::plus<double>(), std::divides<double>());
+    auto rmmw_o = std::inner_product(oxComp, oxComp+m_kk, molecularWeights().data(), 0.0,
+            std::plus<double>(), std::divides<double>());
```

Comment:
I think I would prefer the readability of the straightforward loop for doing this calculation.